libtransmission/torrent.cc

```diff
@@ -2659,6 +2662,13 @@ void tr_torrent::ResumeHelper::load_incomplete_dir(std::string_view const dir) n
 
 // ---
 
+void tr_torrent::ResumeHelper::load_queue_position(size_t pos) noexcept
+{
+    tor_.queue_position_ = pos;
```

Comment:
Potentially problematic. AFAIK we don't have any guarantee that there's not already a torrent with this queue position.

Something like `tr_torrent::set_unique_queue_position()` would be what we want here, but that introduces a computationally expensive cost on startup when loading a number of torrents.

One approach to fix this that I started (but never finished) was to move queue state out of `tr_torrent` into a new Queue class that mapped queue position -> torrent. That way we'd have a single-purpose class to manage the queue and also shave a few bytes off the sizeof `tr_torrent`.

A smaller refactor would be to keep the fields as-is and to allow collisions to occur here, but to add some kind of session-wide idle work a la `tr_torrent::do_idle_work()` to normalize the queue position indices after everything is loaded.

Reply:
> Potentially problematic. AFAIK we don't have any guarantee that there's not already a torrent with this queue position.

Assuming the the resume files were only written by Transmission, then this is not a problem. The resume files on disk should have unique queue numbers, and newly added torrents won't have a resume file.

But this is a big assumption to make, I am looking for more fool-proof ways too.

> One approach to fix this that I started (but never finished) was to move queue state out of `tr_torrent` into a new Queue class that mapped queue position -> torrent.

Hmm, I'm not too sure what you mean here, where can I read those WIP code that you wrote?

If I had to guess, what you mean would be something like a vector of `tr_torrent` pointers, ordered by queue order.

Reply:
> > Potentially problematic. AFAIK we don't have any guarantee that there's not already a torrent with this queue position.
> 
> Assuming the the resume files were only written by Transmission, then this is not a problem. The resume files on disk should have unique queue numbers, and newly added torrents won't have a resume file.
> 
> But this is a big assumption to make, I am looking for more fool-proof ways too.

I think making this assumption is the right way to go, but check for duplicates/holes anyway. (Order torrents by queue position (low->high), iterate, when encountering hole: lower all next torrents queue positions by delta hole size, when encountering duplicates:raise all next torrents queue positions by 1.)

Also, currently the default queue position is derived from id, which is derived from reading the torrent(/resume?) files in order of torrent hash(/name?). Why, when queue positions were introduced, was that not changed to read them in order of file creation date?

Reply:
> I think making this assumption is the right way to go, but check for duplicates/holes anyway.

Aren't making the assumption and check for duplicates/holes literally mutally exclusive...?

> Why, when queue positions were introduced, was that not changed to read them in order of file creation date?

I don't see what difference that makes, and I won't continue this discussion any further in this PR. Not only is it off topic, it is also fruitless.

Reply:
I agree it's not a realistic case unless someone has tempered with the resume data outside of Transmission.
Yet, at Transmission startup we'd probably want to avoid a crash/corruption if that were to happen. So Charles refactoring suggestion is probably desired, but it can certainly be done in a separate pull request instead of this one?

Reply:
Agreed that this should be in a separate PR, as it's getting away from the core of this bugfix & feature. Hopefully should be good to merge?